vsintegration/src/FSharp.Editor/LanguageService/SingleFileWorkspaceMap.fs

```diff
@@ -31,36 +33,91 @@ type internal SingleFileWorkspaceMap(workspace: VisualStudioWorkspace,
+                                match files.TryGetValue(filePath) with
+                                | true, (depProjectContext, _) ->
+                                    projectContext.AddProjectReference(depProjectContext, MetadataReferenceProperties.Assembly)
+                                | _ ->
+                                    let result = createProjectContext filePath
+                                    files.[filePath] <- result
+                                    let depProjectContext, _ = result
+                                    projectContext.AddProjectReference(depProjectContext, MetadataReferenceProperties.Assembly)
+                            )
+
+                    | _ -> ()
+            )
+        )
+
         miscFilesWorkspace.DocumentOpened.Add(fun args ->
             let document = args.Document
 
             if document.Project.Language = FSharpConstants.FSharpLanguageName && workspace.CurrentSolution.GetDocumentIdsWithFilePath(document.FilePath).Length = 0 then
-                files.[document.FilePath] <- createProjectContext document.FilePath
+                let filePath = document.FilePath
+                lock gate (fun () ->
+                    if files.ContainsKey(filePath) |> not then
+                        files.[filePath] <- createProjectContext filePath
+                )
         )
 
         workspace.DocumentOpened.Add(fun args ->
             let document = args.Document
             if document.Project.Language = FSharpConstants.FSharpLanguageName && 
                not document.Project.IsFSharpMiscellaneousOrMetadata then
-                match files.TryRemove(document.FilePath) with
-                | true, projectContext ->
-                    optionsManager.ClearSingleFileOptionsCache(document.Id)
-                    projectContext.Dispose()
-                | _ -> ()
+                optionsManager.ClearSingleFileOptionsCache(document.Id)
+
+                lock gate (fun () ->
+                    match files.TryRemove(document.FilePath) with
+                    | true, (projectContext, _) ->
+                        projectContext.Dispose()
+                    | _ -> ()
+                )
         )
 
         workspace.DocumentClosed.Add(fun args ->
             let document = args.Document
             if document.Project.Language = FSharpConstants.FSharpLanguageName && 
                document.Project.IsFSharpMiscellaneousOrMetadata then
-                match files.TryRemove(document.FilePath) with
-                | true, projectContext ->
-                    optionsManager.ClearSingleFileOptionsCache(document.Id)
-                    projectContext.Dispose()
-                | _ -> ()
+                optionsManager.ClearSingleFileOptionsCache(document.Id)
+
+                lock gate (fun () ->
+                    match files.TryRemove(document.FilePath) with
+                    | true, (projectContext, _) ->
+                        projectContext.Dispose()
```

Comment:
We should not dispose of a `projectContext` if it's being used as a dependency for another project.

Reply:
Note this TODO